server/src/instant/db/model/triple.clj

```diff
@@ -123,152 +124,234 @@
+
+(defn validate-required!
+  "Given [{:entity_id ..., :attr_id ...} ...] of entity attributes that were
+   added or removed during this tx, checks that all affected entities that are still
+   alive have all required attributes set"
+  [conn app-id eid+attr-ids]
+  (let [eid+attr-ids-cte
+        (sql/recordset eid+attr-ids
+                       {'entity_id {:type :uuid, :as 'entity-id}
+                        'attr_id   {:type :uuid, :as 'attr-id}})
+
+        eid+etypes
+        {:from 'eid+attr-ids-cte
+         :join ['attrs  [:= 'attr-id 'attrs/id]
+                'idents [:= 'attrs/forward-ident 'idents/id]]
+         :select-distinct ['entity-id 'etype]}
+
+        eid+required-attrs
+        {:from 'eid+etypes
+         :join ['idents [:and
+                         [:= 'idents/app-id app-id]
+                         [:= 'idents/etype 'idents/etype]]
+                'attrs [:= 'idents/id 'attrs/forward-ident]]
+         :where [:or
+                 [:= 'attrs/is_required true]
+                 [:= 'attrs/setting_required true]]
+         :select-distinct ['entity-id ['attrs/id 'attr-id] 'idents/etype 'idents/label]}
+
+        missing-required
+        {:from      'eid+required-attrs
+         :left-join ['triples
+                     [:and
+                      [:= 'eid+required-attrs/entity-id 'triples/entity-id]
+                      [:= 'eid+required-attrs/attr-id 'triples/attr-id]]]
+         :where     [:= 'triples/value nil]
+         :select    ['eid+required-attrs/* 'triples/value]}
+
+        missing-required-alive
+        {:from ['missing-required
+                [[:lateral {:from 'idents
+                            :where [:and
+                                    [:= 'idents/app-id app-id]
+                                    [:= 'idents/etype 'missing-required/etype]]
+                            :join ['attrs   [:= 'idents/id 'attrs/forward-ident]
+                                   'triples [:= 'triples/attr-id 'attrs/id]]
+                            :select [[[:count 'triples/entity-id] 'cnt]]}] 'cnt]]
+         :where [:> 'cnt/cnt [:inline 0]]
+         :select 'missing-required/*}
+
+        query
+        {:with   [['eid+attr-ids-cte       eid+attr-ids-cte]
+                  ['eid+etypes             eid+etypes]
+                  ['eid+required-attrs     eid+required-attrs]
+                  ['missing-required       missing-required]
+                  ['missing-required-alive missing-required-alive]]
+         :select ['entity-id 'attr-id 'etype 'label]
+         :from   'missing-required-alive}
+
+        res (sql/do-execute! conn (hsql/format query))]
+    (when (seq res)
+      (ex/throw-validation-err!
+       :required-attrs
+       res
+       (for [{entity-id :entity_id
+              etype     :idents/etype
+              label     :idents/label} res]
+         {:message (str (str/capitalize etype) " " entity-id " is missing required attribute '" label "'")})))))
```

Comment:
It might be weird to see the etype capitalized in the error message, I would just leave it lowercase.

Would also be handy to put the `etype`,`entity-id`, and `label` in the hint field.